Sprint 2/step 2

AlexSarva opened the pull request:
Done sprint 2

AlexSarva committed:
feat: add first reducer and add preloader

AlexSarva committed:
feat: add all reducers and navigation

AlexSarva committed:
feat: done some drag-n-drop

AlexSarva committed:
feat: done drag-n-drop

vadmas commented on src/components/app/app.jsx:
```diff
@@ -1,144 +1,63 @@
+      </DndProvider>
 
-      {isIngModalOpen && ingredientInfo && (
+      {showDetails && itemDetails && (
```
Нужно исправить: не стоит частную логику выносить в общий компонент App. Уместнее данное модальное окно вынести в компонент BurgerIngredients

AlexSarva replied:
done

vadmas commented on src/components/app/app.jsx:
```diff
@@ -1,144 +1,63 @@
       )}
-      {isOrderModalOpen && (
+
+      {showOrder && (
```
Нужно исправить: не стоит частную логику выносить в общий компонент App. Уместнее данное модальное окно вынести в компонент BurgerConstructor

AlexSarva replied:
done

vadmas commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -2,116 +2,240 @@ import constructorStyle from './burger-constructor.module.css';
+import {useDrag, useDrop} from "react-dnd";
+import {useRef, useState} from "react";
+
+const ResultInfo = () => {
```
Нужно исправить: компонент стоит вынести в отдельный файл - директорию

AlexSarva replied:
done

vadmas commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -2,116 +2,240 @@ import constructorStyle from './burger-constructor.module.css';
-
-  const deleteIngredient = () => {
-    deleteOption(num, ingredient._id);
+const DragConstructorElement = ({index, ingredient}) => {
```
Нужно исправить: компонент стоит вынести в отдельный файл - директорию (касается всех подобных случаев, когда название компонента не соответствует наименованию файла)

AlexSarva replied:
done

vadmas commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -2,116 +2,240 @@ import constructorStyle from './burger-constructor.module.css';
+  }
+);
+
+  const [{ isOver}, dropPicked] = useDrop({
```
Отлично: хорошо реализована работа с react-dnd

vadmas commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -2,116 +2,240 @@ import constructorStyle from './burger-constructor.module.css';
+      }
+
+      dispatch(setDropIndex({
+        dropIndex: hoverClientY < hoverMiddleY ? dropIndex : (dropIndex + 1),
```
Отлично: правильно, что передаются только индексы перемещаемого элемента

vadmas commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -2,116 +2,240 @@ import constructorStyle from './burger-constructor.module.css';
 }
 
 DragConstructorElement.propTypes = {
   ingredient: ConstructorElementType,
```
Нужно исправить: не хватает isRequired

AlexSarva replied:
done

vadmas commented on src/services/reducers/burger-constructor.jsx:
```diff
@@ -0,0 +1,97 @@
+  initialState: {
+    bun: null,
+    options: [],
+    totalPrice: 0,
```
Можно лучше: полную стоимость лучше рассчитывать в селекторе (на стороне компонента), а не хранить в сторе. В данном случае можно использовать два способа вычисления на стороне компонента: useMemo (мемоизация вычислений на стороне компонента), либо используя библиотку reselect (мемоизация вычислений в хуке useSelect)

vadmas commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,24 +1,39 @@
+
   const handleClick = (e) => {
     onPickCategory(e);
     refs[e].current.scrollIntoView({behavior: "smooth"});
```
Отлично: реализована навигация по разделам ингредиентов - молодец!

vadmas commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -31,116 +46,167 @@ IngredientsNavigation.propTypes = {
 }
 
 Ingredient.propTypes = {
   ingredient: PropTypes.object.isRequired,
```
Надо исправить: по чеклисту требуется полное описание структуры объектов, а не просто object

AlexSarva replied:
done

vadmas commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -31,116 +46,167 @@ IngredientsNavigation.propTypes = {
 }
 
-const Ingredients = forwardRef(({title, ingredients, onAddClick, ingredientsCount, onClickIng}, ref) => {
+const Ingredients = forwardRef(({type}, ref) => {
```
Отлично: хороший кейс для использования forwardRef

vadmas commented on src/services/reducers/ingredient-info.jsx:
```diff
@@ -0,0 +1,23 @@
+const ingredientInfoSlice = createSlice({
+  name: 'ingredientInfo',
+  initialState: {
+    itemDetails: {},
```
Нужно исправить: отсутствие данных правильно обозначать как null а не пустой объект. Соответственно там, где запрашиваются свойства данного объекта в компоненте, следует сделать проверку на null, добавив тернарный оператор (object?.property)

AlexSarva replied:
done

vadmas commented on src/services/reducers/ingredients.jsx:
```diff
@@ -0,0 +1,83 @@
+        if (action.payload.success === true ) {
+          state.items.bun = action.payload.data.filter(item => item.type === 'bun')
+          state.items.main = action.payload.data.filter(item => item.type === 'main')
+          state.items.sauce = action.payload.data.filter(item => item.type === 'sauce')
```
Можно лучше: переиспользуемые строковые значения типа bun, main, sauses – лучше выносить в константы

AlexSarva replied:
done

vadmas commented on src/services/reducers/orders.jsx:
```diff
@@ -0,0 +1,62 @@
+    },
+    status: 'idle',
+    error: null,
+    showOrder: false,
```
Нужно исправить: данный параметр избыточен. Модальное окно следует открывать, если данные в сторе есть (order.number).

AlexSarva replied:
Использую showOrder исключительно для того, чтобы показать прелоадер и ошибку в случае ошибке при запросе 

<img width="379" alt="image" src="https://user-images.githubusercontent.com/59559131/235610193-a45122ff-883c-4575-936d-bce255741f88.png">

vadmas replied:
Отлично. Тогда его использование логично - это хорошая доработка! (извиняюсь, сразу не углядел)

vadmas reviewed (commented):
Хорошая работа.
Приложение приобретает готовый вид, видно как вы прогрессируете по мере прохождения курса.
В положительную сторону хочется отметить:
- Хорошо реализована работа с redux-thunk
- Cторы - слайсы, разделены по сущностям - молодец!
- Взаимодействие сторов с компонентами реализовано через хуки useSelector, useDispatch
- Работы с redux-dnd реализована отлично - молодец!
- Реализована навигация по разделам ингредиентов - отлично!
- Функционал конструктора бургера реализован верно
- Подключено расширение redux devtools

**Но есть несколько замечаний:**

**Замечания которые необходимо будет исправить:**
- Модальные окна целесообразно выносить в контейнеры сущностей (смотри комментарии в компоненте App), т.к. хранение их в компоненте App (старший компонент приложения) будет вызывать перерендер всего приложения, при изменении состояний параметров, необходимых для модальных окон.
- Целесообразно называть папки компонентов - одноименным названием файлов-компонентов. Так же стоит придерживаться правила: один компонент - один файл - одна папка (и у всех одноименное название). В данной работе в файлах burger-constructor.jsx, burger-ingredients.jsx - реализовано несколько компонентов

Так же стоит учесть комментарии, оставленные в коде.

**Места, в которых можно улучшить проект:**
- Вычисления и преобразование данных (на стороне компонента), полученных из селектора (хук useSelector), можно делать в селекторе, используя библиотеку reselect. Данная библиотека нужна для создания мемоизированных, пригодных для компоновки селекторных функций. Примеры взаимодействия reselect с useSelector можно посмотреть здесь: https://react-redux.js.org/api/hooks#useselector-examples
- Вычисления, фильтрацию и маппинг данных, выполняемых на стороне компонента, целесообразно оборачивать в хук useMemo. Так повторные пересчеты значений не будут выполняться на каждый перерендер компонента, а только при изменении установленных зависимостей. Подробнее о данном хуке можно почитать здесь: https://ru.reactjs.org/docs/hooks-reference.html#usememo

Пожалуйста, учтите выше описанные замечания и те, что оставлены коде вашей работы (особенно с пометкой "нужно исправить").
При повторной проверке, если все замечания будут исправлены, работа будет принята.

Удачного рефакторинга кода!

AlexSarva committed:
fix: moved modals to components

AlexSarva committed:
fix: refactor burger-ingredients component

AlexSarva committed:
feat(burger-constructor): добавлены новые компоненты

Добавлены новые компоненты burger-components, drag-constructor-element, empty-constructor, result-info и соответствующие им файлы стилей. Эти компоненты будут использоваться в конструкторе бургеров для отображения информации о добавленных ингредиентах, перетаскивания элементов, пустого конструктора и информации о заказе.

AlexSarva committed:
fix: refactor burger-ingredients component

AlexSarva committed:
feat: добавлены константы для типов ингредиентов в файле constants.js
fix: исправлены места использования строковых литералов на константы в файлах burger-components.jsx, drag-constructor-element.jsx и ingredients.jsx
refactor: изменены фильтры в reducers/ingredients.jsx на использование констант из файла constants.js
Добавлены константы для типов ингредиентов в файле constants.js, что улучшает читаемость кода и уменьшает вероятность ошибок при использовании строковых литералов. Исправлены места использования строковых литералов на константы в файлах burger-components.jsx, drag-constructor-element.jsx и ingredients.jsx. В reducers/ingredients.jsx изменены фильтры на использов

vadmas commented:
Отлично, все критичные комментарии исправлены. Работу принимаю. Удачи в дальнейшем обучении!
